code/datums/components/payment.dm

```diff
@@ -69,7 +73,34 @@
 
 /datum/component/payment/proc/change_cost(datum/source, new_cost)
 	SIGNAL_HANDLER
-
 	if(!isnum(new_cost))
 		CRASH("change_cost called with variable new_cost as not a number.")
 	cost = new_cost
+
+/datum/component/payment/proc/purchase_item(datum/source, atom/movable/customer)
+	SIGNAL_HANDLER
+	if(!LAZYLEN(products))
+		return
+	var/list/display_names = list()
+	var/list/items = list()
+	for(var/i in 1 to length(products))
+		var/obj/item/product = products[i]
+		display_names["[initial(product.name)] ([i])"] = REF(product)
+		var/image/product_image = image(icon = initial(product.icon), icon_state = initial(product.icon_state))
+		items += list("[initial(product.name)] ([i])" = product_image)
+	var/pick = show_radial_menu(customer, parent, items, custom_check = FALSE, require_near = TRUE)
+	if(!pick)
+		return
+	var/product_reference = display_names[pick]
+	var/obj/item/new_product = locate(product_reference) in products
+	if(!new_product)
+		return
+	if(attempt_charge(src, customer, initial(new_product.custom_price)) & COMPONENT_OBJ_CANCEL_CHARGE)
+		return
+	new new_product(customer.drop_location())
+	for(var/i in 1 to length(products))
```

Comment:
My brain is almost certainly soft because I've forgotten how to key lists and this part for incrementing bought objects is 100% not incrementing down properly

Reply:
For what you need this `for` if you have `new_product` that already position in `products`
Try
```
		products[new_product] -= 1
		if(products[new_product] < 1)
			products -= new_product
```